### PR TITLE
Set real time priority for SoundDeviceNetworkThread on Linux

### PR DESCRIPTION
**Pull request by daschuer**
p_pThread->start(QThread::TimeCriticalPriority); does not work on Linux.
As reported here https://bugs.launchpad.net/mixxx/+bug/1641359

I have ported the code from Portaudio that is called if you set PaAlsa_EnableRealtimeScheduling

The result is really great, it allows to use 2.67 ms buffer on my Ubuntu Trusty system without a single underflow. 

Soundcard thread and network thread are on the same priority now. See:

```
$ ps -T -C19,_20 -o priority,ni,pcpu,pid,comm 18061
PRI  NI %CPU   PID COMMAND
 20   0 18.9 18061 mixxx
 20   0  0.0 18061 dconf worker
 20   0  0.0 18061 gdbus
 20   0  0.0 18061 gmain
 20   0  1.0 18061 StatsManager
 20   0  0.0 18061 mixxx
 20   0  0.0 18061 EngineWorkerSch
 20   0  0.0 18061 EngineSideChain
 20   0  0.0 18061 VinylControlPro
 20   0  0.1 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 LibraryScanner 
 20   0  0.7 18061 BrowseThread
 20   0  0.4 18061 AnalyzerQueue
 20   0  0.7 18061 Controller
 20   0  1.0 18061 VSyncThread
 20   0  0.0 18061 QProcessManager
 20   0  0.0 18061 Controller
 -2   -  1.6 18061 mixxx
 -2   -  8.1 18061 SoundDeviceNetw
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo
 20   0  0.0 18061 CachingReaderWo

```

### CODE REVIEW

**Comment by Be-ing** [comment]
> Soundcard thread and network thread are on the same priority now.

This should only happen if they are both in use. If the network clock is being used without any sound card output, the sound card thread should not have priority. Likewise, if the network clock isn't being used, it shouldn't get priority.

**Comment by daschuer** [comment]
>>  Sundcard thread and network thread are on the same priority now.

> This should only happen if they are both in use. If the network clock is being used without any sound card output, the sound card thread should not have priority. Likewise, if the network clock isn't being used, it shouldn't get priority.

If one of these threads is not used it does not even exist.

File: src/soundio/sounddevicenetwork.h [hunk 1]
```diff
@@ -85,6 +89,16 @@ class SoundDeviceNetworkThread : public QThread {
 
   private:
     void run() {
+
+#ifdef __LINUX__
```
Comment (Be-ing) [inline src/soundio/sounddevicenetwork.h]: Please add a comment for why this is only necessary on Linux.

File: src/soundio/sounddevicenetwork.h [hunk 2]
```diff
@@ -85,6 +89,16 @@ class SoundDeviceNetworkThread : public QThread {
...
+        struct sched_param spm = { 0 };
+        spm.sched_priority = 1;
+        if(pthread_setschedparam(pthread_self(), SCHED_FIFO, &spm)
+        {
```
Comment (Be-ing) [inline src/soundio/sounddevicenetwork.h]: `{` should be on line above

**Review by Be-ing — APPROVED** [review]
Good idea with this. Please add a comment in the code explaining why the change was made.

File: src/soundio/sounddevicenetwork.h (hunk 1 above)
Comment (daschuer) [reply src/soundio/sounddevicenetwork.h]: I have no idea how the situation is on the other OSs alt least portaudio has this feature for ALSA only. 
Who can verify this on windows an Mac Os?
Is there a tool to visualise thread priorities on windows?

**Comment by poelzi** [comment]
I started working on this in  #1382 as well. I added a Helper Class for this that also tries realtimekit if the user has no realtime priority limits. Unfortunatelly I came to the conclusion that this will not work as it requires a rtime limit be set and this will cause the audio thread to get killed by the kernel on high load. 2 rubberbands on and you get killed quite early.

**Comment by daschuer** [comment]
I will merge it now without a formal LGTM, because this should be stressed during beta period. 
Remaining comments can be fixed afterwards.

**Pull request merged**